napari/viewer.py

```diff
@@ -58,6 +58,8 @@ def __init__(
         logopath = join(dirname(__file__), 'resources', 'logo.png')
         app.setWindowIcon(QIcon(logopath))
 
+        app.aboutToQuit.connect(wait_for_workers_to_quit)
```

Comment:
Does wait actually wait, or does it kill them after the timeout? I wonder if there's a better name for it, say, `shut_down_workers_by_hook_or_by_crook` or similar. :stuck_out_tongue:

Reply:
Short answer: the default behavior of `wait_for_workers_to_quit` is to request all threads to quit, then block _indefinitely_ until threads finish (the safest... leaving it up to the user to "force quit" if a thread is not gracefully closing).  You can call it with a timeout `wait_for_workers_to_quit(1000)` that will return False if 1 second passes and threads have still not successfully finished... and we could decide to try to kill threads ourselves at that point.

But this is another super interesting and important detail (and I'm definitely still learning!).  Turns out it is _extremely_ hard to kill a misbehaving thread that is not continually "asking" whether it should be killed.  ([This SO thread](https://stackoverflow.com/questions/323972/is-there-any-way-to-kill-a-thread) is super useful).  So that is one reason why the `start_worker` function demands that the worker be an instance of `WorkerBase` ... where we know we've added some periodic checks for `_abort_requested` in the event loop.  But here too, there is a catch here (which again, I will add to the user-facing docs).  If the user does _not_ put any yields in their function, and the function is super long, it will just hang...  For instance:
```python
@thread_worker
def ZZZzzz():
    time.sleep(10000000)
```

There's no graceful way to kill this thread in python.  This is why I added the yield to the `layer_update` function in `examples/interactive_scripting.py`.  My (preliminary) understanding from [that SO post](https://stackoverflow.com/a/325528/1631624) is that, should we choose to, we can might be able to use `ctypes` to leverage OS-specific mechanisms to forcefully (and dangerously) kill misbehaving threads.  But if I'm not mistaken, that's not dissimilar from the user having to force quit... and I opted (for now) to leave that last `by_hook_or_by_crook` step to the user...

Curious what you think.  I agree from a user perspective, if you say quit, you want **QUIT**.  I just don't yet know enough about this stuff to feel super comfortable baking in potentially dangerous thread-killing logic.  (maybe a future PR?)

Reply:
Awesome, thank you for writing this, it's super useful. As long as this makes it into the docs in some form, I'm happy to postpone the decision on whether to go on a murderous thread rampage.

Reply:
added to docs